### PR TITLE
Removing Generate*() methods.

### PR DESCRIPTION
**Pull request by sornaks**
1. Removing ITextWriterCopyable - since we made TagHelperContent enumerable.
2. Removing HtmlEncoder - since generating start tag logic is moved to MVC.Razor.
3. Removing Generate*() methods and corresponding tests.
4. Introducing WriteTagHelper() method.

https://github.com/aspnet/Razor/issues/318

### CODE REVIEW

File: src/Microsoft.AspNet.Razor.Runtime/TagHelpers/TagHelperOutput.cs [hunk 1]
```diff
@@ -33,19 +30,15 @@ internal TagHelperOutput(string tagName)
         /// </summary>
         /// <param name="tagName">The HTML element's tag name.</param>
         /// <param name="attributes">The HTML attributes.</param>
-        /// <param name="htmlEncoder">The <see cref="IHtmlEncoder"/> used
-        /// to encode HTML attribute values.</param>
         public TagHelperOutput(
             string tagName,
-            [NotNull] IDictionary<string, string> attributes,
-            [NotNull] IHtmlEncoder htmlEncoder)
```
Comment (NTaylorMullen) [inline src/Microsoft.AspNet.Razor.Runtime/TagHelpers/TagHelperOutput.cs]: Ahhh awesome no longer need this. 

You should also remove this: https://github.com/aspnet/Razor/blob/dev/src/Microsoft.AspNet.Razor.Runtime/project.json#L9

File: src/Microsoft.AspNet.Razor/Generator/Compiler/CodeBuilder/CSharp/CSharpTagHelperCodeRenderer.cs [hunk 2]
```diff
@@ -432,25 +359,27 @@ private void RenderEndTagHelpersScope()
...
         }
 
-        private void RenderTagOutput(string tagOutputMethodName, bool methodReturnsString)
+        private void RenderWriteTagHelper()
+        {
+            RenderStartWriteTagHelperMethod();
+            _writer.Write(ExecutionContextVariableName).WriteEndMethodInvocation();
+        }
+
+        private void RenderStartWriteTagHelperMethod()
```
Comment (dougbu) [inline src/Microsoft.AspNet.Razor/Generator/Compiler/CodeBuilder/CSharp/CSharpTagHelperCodeRenderer.cs]: nit: not used elsewhere.  no need to separate from `RenderWriteTagHelper()`

File: src/Microsoft.AspNet.Razor/Generator/GeneratedClassContext.cs [hunk 3]
```diff
@@ -65,33 +70,41 @@ public struct GeneratedClassContext
         public GeneratedClassContext(string executeMethodName,
                                      string writeMethodName,
                                      string writeLiteralMethodName,
+                                     string writeTagHelperMethodName,
                                      string writeToMethodName,
                                      string writeLiteralToMethodName,
+                                     string writeTagHelperToMethodName,
```
Comment (dougbu) [inline src/Microsoft.AspNet.Razor/Generator/GeneratedClassContext.cs]: suggest these should be part of `GeneratedTagHelperContext` instead

File: src/Microsoft.AspNet.Razor/Generator/GeneratedClassContext.cs [hunk 4]
```diff
@@ -124,8 +141,10 @@ public struct GeneratedClassContext
         // Required Items
         public string WriteMethodName { get; private set; }
         public string WriteLiteralMethodName { get; private set; }
+        public string WriteTagHelperMethodName { get; private set; }
         public string WriteToMethodName { get; private set; }
         public string WriteLiteralToMethodName { get; private set; }
+        public string WriteTagHelperToMethodName { get; private set; }
         public string ExecuteMethodName { get; private set; }
         public GeneratedTagHelperContext GeneratedTagHelperContext { get; private set; }
```
Comment (dougbu) [inline src/Microsoft.AspNet.Razor/Generator/GeneratedClassContext.cs]: nits:
- unless these are serialized somewhere along the line (@NTaylorMullen ?), can remove the private setters
- might also be worth adding blank lines between these properties, giving them some room to breath / be read

File: src/Microsoft.AspNet.Razor/Generator/GeneratedClassContext.cs [hunk 5]
```diff
@@ -124,8 +141,10 @@ public struct GeneratedClassContext
         // Required Items
         public string WriteMethodName { get; private set; }
         public string WriteLiteralMethodName { get; private set; }
+        public string WriteTagHelperMethodName { get; private set; }
```
Comment (dougbu) [inline src/Microsoft.AspNet.Razor/Generator/GeneratedClassContext.cs]: docs for new properties

File: test/Microsoft.AspNet.Razor.Runtime.Test/TagHelpers/TagHelperOutputTest.cs [hunk 6]
```diff
@@ -33,329 +32,42 @@ public void TagName_CanSetToNull()
...
-            // Act
-            var output = tagHelperOutput.GenerateStartTag();
-
-            // Assert
-            Assert.Equal("<p class=\"btn\" something=\"   spaced    \">", output);
-        }
-
-        [Fact]
-        public void GenerateStartTag_ReturnsNoAttributeStartTag()
+        public void CanSetPreContent()
```
Comment (dougbu) [inline test/Microsoft.AspNet.Razor.Runtime.Test/TagHelpers/TagHelperOutputTest.cs]: - `PreContent_SetContent_ChangesValue` and same for other new tests.  ("can" isn't an expectation and it should not appear in test names.)
- better to confirm all `TagHelperContent` properties are non-`null` in a newly-initialized `TagHelperOutput` instance.  these are currently `TagHelperContent` tests that assume that is true.

File: test/Microsoft.AspNet.Razor.Runtime.Test/TagHelpers/TagHelperOutputTest.cs [hunk 7]
```diff
@@ -33,329 +32,42 @@ public void TagName_CanSetToNull()
...
-            var tagHelperOutput = new TagHelperOutput("p");
-            tagHelperOutput.PreContent.Append("Hello World");
-
-            // Act
-            var output = tagHelperOutput.GeneratePreContent();
-
-            // Assert
-            var result = Assert.IsType<DefaultTagHelperContent>(output);
+            // Act & Assert
+            var result = Assert.IsType<DefaultTagHelperContent>(tagHelperOutput.PreContent);
```
Comment (dougbu) [inline test/Microsoft.AspNet.Razor.Runtime.Test/TagHelpers/TagHelperOutputTest.cs]: this assertion confirms a low-level detail (the implementation type).  should instead `Assert.NotNull(tagHelperOutput.PreContent)`

File: src/Microsoft.AspNet.Razor/Generator/Compiler/CodeBuilder/CSharp/CSharpTagHelperCodeRenderer.cs [hunk 8]
```diff
@@ -432,25 +359,27 @@ private void RenderEndTagHelpersScope()
...
-
-            _writer.Write(ExecutionContextVariableName)
-                   .Write(".")
-                   .Write(_tagHelperContext.ExecutionContextOutputPropertyName)
-                   .Write(".")
-                   .WriteMethodInvocation(tagOutputMethodName, endLine: false)
-                   .WriteEndMethodInvocation();
         }
+
         private void RenderStartWriteMethod()
```
Comment (NTaylorMullen) [inline src/Microsoft.AspNet.Razor/Generator/Compiler/CodeBuilder/CSharp/CSharpTagHelperCodeRenderer.cs]: This method is no longer used, can remove it.

File: test/Microsoft.AspNet.Razor.Test/TestFiles/CodeGenerator/CS/Output/BasicTagHelpers.CustomAttributeCodeGenerator.cs [hunk 9]
```diff
@@ -36,22 +36,7 @@ public BasicTagHelpers()
...
-                {
-                    Write(__tagHelperExecutionContext.GetChildContentAsync().Result);
-                }
-                else
-                {
-                    __tagHelperExecutionContext.ExecuteChildContentAsync().Wait();
-                }
-                Write(__tagHelperExecutionContext.Output.GeneratePostContent());
-                WriteLiteral(__tagHelperExecutionContext.Output.GenerateEndTag());
+                WriteTagHelper(__tagHelperExecutionContext);
```
Comment (dougbu) [inline test/Microsoft.AspNet.Razor.Test/TestFiles/CodeGenerator/CS/Output/BasicTagHelpers.CustomAttributeCodeGenerator.cs]: :dancers:

File: src/Microsoft.AspNet.Razor/Generator/GeneratedClassContext.cs (hunk 3 above)
Comment (NTaylorMullen) [inline src/Microsoft.AspNet.Razor/Generator/GeneratedClassContext.cs]: :+1:

**Comment by dougbu** [comment]
:watch:

File: src/Microsoft.AspNet.Razor/Generator/GeneratedClassContext.cs (hunk 4 above)
Comment (NTaylorMullen) [inline src/Microsoft.AspNet.Razor/Generator/GeneratedClassContext.cs]: Not serialized anywhere

**Comment by NTaylorMullen** [comment]
:watch: only for the config changes.

**Comment by sornaks** [comment]
Addressed comments.

**Comment by NTaylorMullen** [comment]
:shipit: once we sign of on the MVC PR

**Comment by NTaylorMullen** [comment]
:watch: https://github.com/aspnet/Mvc/pull/2180/files#r26512936

File: src/Microsoft.AspNet.Razor/Generator/Compiler/CodeBuilder/CSharp/CSharpTagHelperCodeRenderer.cs [hunk 10]
```diff
@@ -75,15 +75,8 @@ public void RenderTagHelper(TagHelperChunk chunk)
...
-                RenderTagOutput(_tagHelperContext.OutputGenerateStartTagMethodName, methodReturnsString: true);
-                RenderTagOutput(_tagHelperContext.OutputGeneratePreContentMethodName, methodReturnsString: false);
-
-                RenderTagHelperContent();
-
-                RenderTagOutput(_tagHelperContext.OutputGeneratePostContentMethodName, methodReturnsString: false);
-                RenderTagOutput(_tagHelperContext.OutputGenerateEndTagMethodName, methodReturnsString: true);
-
+                RenderStartWriteTagHelperMethod();
+                _writer.Write(ExecutionContextVariableName).WriteEndMethodInvocation();
```
Comment (dougbu) [inline src/Microsoft.AspNet.Razor/Generator/Compiler/CodeBuilder/CSharp/CSharpTagHelperCodeRenderer.cs]: nit: couldn't this be done inside `RenderStartWriteTagHelperMethod()`, perhaps after renaming that method to (say) `RenderWriteTagHelperMethodCall`?

File: src/Microsoft.AspNet.Razor/Generator/GeneratedTagHelperContext.cs [hunk 11]
```diff
@@ -154,16 +100,21 @@ public GeneratedTagHelperContext()
...
         /// </remarks>
         public string TagHelperContentTypeName { get; set; }
+
+        /// <summary>
+        /// The Write method name for tag helpers.
+        /// </summary>
+        public string WriteTagHelperMethodName { get; set; }
+
+        /// <summary>
+        /// The WriteTo method name for tag helpers.
```
Comment (dougbu) [inline src/Microsoft.AspNet.Razor/Generator/GeneratedTagHelperContext.cs]: rewrite these comments to be consistent w/ the rest of those in the class e.g. `The name of the method used to write <see cref="ExecutionContextTypeName"/> content.`

**Comment by dougbu** [comment]
:watch: please grab me when you have updated the new doc comments

**Comment by sornaks** [comment]
Updated..

File: src/Microsoft.AspNet.Razor/Generator/Compiler/CodeBuilder/CSharp/CSharpTagHelperCodeRenderer.cs [hunk 12]
```diff
@@ -432,38 +359,23 @@ private void RenderEndTagHelpersScope()
...
             }
             else
             {
-                _writer.WriteStartMethodInvocation(_context.Host.GeneratedClassContext.WriteMethodName);
+                _writer.WriteStartMethodInvocation(_tagHelperContext.WriteTagHelperAsyncMethodName);
             }
+
+            _writer.Write(ExecutionContextVariableName)
+                   .WriteEndMethodInvocation(endLine: false)
+                   .WriteLine(".Wait();");
```
Comment (dougbu) [inline src/Microsoft.AspNet.Razor/Generator/Compiler/CodeBuilder/CSharp/CSharpTagHelperCodeRenderer.cs]: teensy nit: here's where my commit comment lands.  you're indenting the easy and easy-to-maintain way just above (line 366) but doing it the hard way here.

**Comment by dougbu** [comment]
:shipit:

**Comment by NTaylorMullen** [comment]
:shipit: once we sign off on MVC PR

**Comment by sornaks** [comment]
Checked in. Thanks! https://github.com/aspnet/Razor/commit/7daec14e49b96ff7833409c465ec73ce9b151e8b

**Pull request closed without merging**